netmiko/fortinet/fortinet_ssh.py

```diff
@@ -1,104 +1,298 @@
+		)
+
+	def exit_config_mode(self, exit_config="end", pattern=r"(#|\$)"):
+		"""Exit from configuration mode.
+
+		:param exit_config: Command to exit configuration mode
+		:type exit_config: str
+
+		:param pattern: Pattern to terminate reading of channel
+		:type pattern: str
+		"""
+		return super().exit_config_mode(
+			exit_config=exit_config, pattern=pattern
+		)
+
+	def disable_paging(
+		self,
+		command: str = "",
+		delay_factor: Optional[float] = None,
+		cmd_verify: bool = True,
+		pattern: Optional[str] = None,
+		) -> str:
+		"""Disable paging is only available with specific roles so it may fail.
+
+		:param command: Device command to disable pagination of output
+
+		:param delay_factor: Deprecated in Netmiko 4.x. Will be eliminated in Netmiko 5.
+		"""
+
+		disable_paging_commands = [
+			"config global",
+			"config system console",
+			"set output standard",
+			"end"
+			]
+		output = self.send_config_set(
+			config_commands = disable_paging_commands,
+			exit_config_mode = True,
+			read_timeout = None,
+			delay_factor = 1.0,
+			max_loops = 150,
+			enter_config_mode = False,
+		)
+		log.debug("***Window Paging Disabled***")
+		return output
+
+	def cleanup(self, command: str = "quit") -> None:
+		"""Re-enable paging globally."""
+		# Return paging state
+		enable_paging_commands = ["config global", "config system console", "set output more", "end"]
+		# Should test output is valid
+		output = self.send_config_set(
+			config_commands = enable_paging_commands,
+			exit_config_mode = True,
+			read_timeout = None,
+			delay_factor = 1.0,
+			max_loops = 150,
+			enter_config_mode = False,
+		)
+		log.debug("***Window Paging Enabled***")
+
+		"""Gracefully exit the SSH session."""
+		try:
+			self.exit_config_mode()
+		except Exception:
+			pass
+		# Always try to send final 'exit' (command)
+		self._session_log_fin = True
+		self.write_channel(command + self.RETURN)
+
+	def save_config(
+		self, cmd: str = "", confirm: bool = False, confirm_response: str = ""
+		) -> str:
+		"""Not Implemented, platform automatically saves"""
+		raise NotImplementedError
+
+	def _enter_global_or_vdom(self, mode_name: str = None):
+		"""This allows a device enter global or vdom mode and confirm before proceeding.
+
+		:param mode_name: the name of the mode trying to be entered: 'global', 'root' or the name of the vdom string.
+		"""
+		if mode_name == "global":
+			log.debug("***Entering Global Mode***")
+			try: 
+				self.send_command(command_string="config global", expect_string=r"\(global\) # ")
+			except:
+				raise ValueError("Device Failed to Enter Global Mode.")
+		else:
+			log.debug("***Entering VDOM Mode***")
+			try:
+				self.send_command(command_string="config vdom", expect_string=r"\(vdom\) # ")
+				self.send_command(command_string=f"edit {mode_name:}", expect_string=fr"\({mode_name}\) # ")
+			except:
+				raise ValueError(f"Device Failed to Enter '{mode_name}' VDOM Mode.")
```

Comment:
When is this used? We should probably decouple these two items and create two separate methods (or just one if only one is needed).

Reply:
Well, they're need for multi-vdom mode if you don't want the user to define it in the "send_config_set" commands as I explained above.  If the device is single-vdom mode, then these methods aren't needed, it's contextual.  I've personally always handle entering and exiting the vdom like a configuration mode and do it in my "send_config_set" script.  I only developed this so it could pass the pytest scripts.

Reply:
Getting more maybe at your intended point of separating them.  The global mode becomes it's own thing like a vdom, it's difficult to explain to someone if they haven't used a Fortigate, but here's my attempt.

If you wanted to change something for snmp for single-vdom, everything is contained with root, by default so it's pretty simple:
"config system snmp user"

However, if multi-vdom mode is active, then you then sets certain configuration under "global" and some under "root" and then some under whatever additional vdom's you created like "management".  Global configs apply to all vdoms and are mostly system related configs.  Root is also device management related, but nothing in root has a relationship to another vdom.  There's almost like a VRF, but more like a Virtual Domain (hence vdom), it has it's own routing table and firewall policies.  Like different customers for example.

Now, I included them in the same Method because they are similar as you can see, but it makes it easier in my opinion to have 1 dedicated Method to enter into global or vdom, less commands to type overall.  so you could type _enter_global_or_vdom("global") or you could type _enter_global_or_vdom("root").   If you want to separate them, I'm fine with that too, was just my original thought process.

Reply:
I will propose an alternate solution for this. I am pretty sure we can create an alternate solution for the test code problem which will be more maintainable in the long run.

Reply:
Have you come up with a solution for this yet or does my proposed solution work?

Reply:
Yes, I will fix. I haven't done it yet, but I will.